bin/fit_posparams

```diff
@@ -0,0 +1,383 @@
+    key_search = [key in frame.columns for key in required_keys]
+    if all(key_search):
+        frames[posid] = frame
+posids = sorted(csv_files.keys())
+
+# load petal alignments
+
+petal_alignments_file = 'petal-alignments.yaml'
+petal_alignments_path = os.path.join(desimeter_data_dir, petal_alignments_file)
+with open(petal_alignments_path, 'r') as file:
+    petal_alignments = yaml.safe_load(file)
+ptl_trans = {}
+for petal_loc, alignment in petal_alignments.items():
+    args = (alignment[key] for key in ['Tx', 'Ty', 'Tz', 'alpha', 'beta', 'gamma'])
+    ptl_trans[petal_loc] = petaltransforms.PetalTransforms(*args, curved=True)
+
+# obtain xy in petal coordinate system
+for posid in frames:
+    petal_loc = frames[posid]['PETAL_LOC'][0]
+    trans = ptl_trans[petal_loc]
+    obsX = frames[posid]['X_FP']
+    obsY = frames[posid]['Y_FP']
+    obsXY = numpy.array([obsX, obsY])
+    ptlXY = trans.obsXY_to_ptlXYZ(obsXY, cast=False)
+    frames[posid]['X_PTL'] = ptlXY[0]
+    frames[posid]['Y_PTL'] = ptlXY[1]
+
+# filter out bad rows
+for posid in frames:
+    # drop any rows with disabled control (not valid data)
+    frames[posid] = frames[posid][frames[posid]['CTRL_ENABLED'] == True]
+    frames[posid] = frames[posid].reset_index(drop=True)
+
+    # drop any rows where no move sequence was performed
+    moveseq_diff = frames[posid]['TOTAL_MOVE_SEQUENCES'].diff()
+    frames[posid] = frames[posid][moveseq_diff != 0]
+    frames[posid] = frames[posid].reset_index(drop=True)
+    
+    # drop any rows where no cruise move was performed
+#    cruiseT_diff = frames[posid]['TOTAL_CRUISE_MOVES_T'].diff()
+#    cruiseP_diff = frames[posid]['TOTAL_CRUISE_MOVES_P'].diff()
+#    frames[posid] = frames[posid][cruiseT_diff + cruiseP_diff != 0]
+#    frames[posid] = frames[posid].reset_index(drop=True)
+
+# add a time-since-epoch column to data frames
+date_format_data = '%Y-%m-%d %H:%M:%S'
+for posid, frame in frames.items():
+    time_structs = [time.strptime(frame['DATE'][i], date_format_data) for i in range(len(frame))]
+    time_sec = [time.mktime(t) for t in time_structs]
+    frame['DATE_SEC'] = time_sec
+
+# data selectors
+def select_by_index(posid, start=0, final=-1):
+    '''Returns a subset of data formatted for the best-fitting function.'''
+    frame = frames[posid]
+    data = {}
+    subframe = frame[start:final]
+    data['posintT'] = subframe['POS_T'].tolist()
+    data['posintP'] = subframe['POS_P'].tolist()
+    data['ptlX'] = subframe['X_PTL'].tolist()
+    data['ptlY'] = subframe['Y_PTL'].tolist()
+    return data, subframe
+
+def first_time_match(frame, time_in_sec):
+    '''Returns index of first row in dataframe which matches the argued time.'''
+    return frame[frame['DATE_SEC'] >= time_in_sec].index[0]
+
+def first_date_in_sec(frame):
+    return frame['DATE_SEC'].tolist()[0]
+
+def last_date_in_sec(frame):
+    return frame['DATE_SEC'].tolist()[-1]
+
+def slice_idxs_for_dates(posid, start=None, final=None):
```

Comment:
Throughout the code, please add info to the docstrings about the input params, not just the return values.  e.g. here I assume that `posid` is the positioner ID string, but I don't know how to interpret `start` and `final` without reading the code more (restricting the indices of the search? is `final` inclusive or not?).  e.g. like you did in the `process` docstring.

Reply:
Sure